### PR TITLE
customizations: add support for fs sizes with custom units

### PR DESCRIPTION
**Pull request by msehnout**
Replaces: #1907

This pull request includes:

- [ ] adequate testing for the new functionality or fixed issue
- [ ] adequate documentation informing people about the change such as
  - [ ] create a file in [news/unreleased](https://github.com/osbuild/osbuild-composer/tree/main/docs/news/unreleased) directory if this change should be mentioned in the release news
  - [ ] submit a PR for the [guides](https://github.com/osbuild/guides) repository if this PR changed any behavior described there: https://www.osbuild.org/guides/

### CODE REVIEW

File: internal/common/helpers.go [hunk 1]
```diff
@@ -36,3 +38,51 @@ func IsStringInSortedSlice(slice []string, s string) bool {
...
+	// Get the number from the string
+	plain_number := regexp.MustCompile(`[[:digit:]]+`)
+	number_as_str := plain_number.FindString(size)
+	if number_as_str == "" {
+		return nil
+	}
+
+	// Parse the number into integer, be aware that the function
+	// returns int64, so it cannot hold all possible values for uint64
+	number, err := strconv.ParseInt(number_as_str, 10, 64)
```
Comment (ondrejbudai) [inline internal/common/helpers.go:54]: cannot `ParseUint` be used?

File: internal/common/helpers.go [hunk 2]
```diff
@@ -36,3 +38,51 @@ func IsStringInSortedSlice(slice []string, s string) bool {
...
+	}
+	return_size := new(uint64)
+	*return_size = uint64(number)
+
+	// List of all supported units (from kB to TB and KiB to TiB)
+	supported_units := []struct {
+		re       *regexp.Regexp
+		multiple uint64
+	}{
+		{regexp.MustCompile(`^\s*[[:digit:]]+\s*kB\s*$`), 1000},
```
Comment (ondrejbudai) [inline internal/common/helpers.go:66]: I suggest calling `strings.TrimSpace` before `MatchString` to simplify the regexes.

File: internal/common/helpers.go [hunk 3]
```diff
@@ -36,3 +38,51 @@ func IsStringInSortedSlice(slice []string, s string) bool {
 	}
 	return false
 }
+
+// DataSizeToUint64 converts a size specified as a string in KB/KiB/MB/etc. to
+// a number of bytes represented by uint64.
+func DataSizeToUint64(size string) *uint64 {
```
Comment (ondrejbudai) [inline internal/common/helpers.go:44]: Returning `(uint64, error)` feels more natural to me but that might be just my opinion. :D

File: internal/blueprint/customizations.go [hunk 4]
```diff
@@ -72,8 +72,30 @@ type ServicesCustomization struct {
 }
 
 type FilesystemCustomization struct {
-	Mountpoint string `json:"mountpoint,omitempty" toml:"mountpoint,omitempty"`
-	MinSize    uint64 `json:"minsize,omitempty" toml:"size,omitempty"`
+	Mountpoint string `json:"mountpoint,omitempty"`
+	MinSize    uint64 `json:"minsize,omitempty"`
+}
+
+func (fsc *FilesystemCustomization) UnmarshalTOML(data interface{}) error {
```
Comment (ondrejbudai) [inline internal/blueprint/customizations.go:81]: We also support JSON so we need a JSON unmarshaller.

Comment (msehnout) [reply internal/blueprint/customizations.go:81]: This is mentioned in the commit message. If this feature is mainly for the interactive usage, it doesn't need to be implemented for JSON. But I'm fine implementing the same for JSON.

**Comment by ondrejbudai** [comment]
I think that base tests are failing for this PR, can you look at them, please?

**Comment by msehnout** [comment]
> I think that base tests are failing for this PR, can you look at them, please?

Fixed ;-)

**Review by ondrejbudai — APPROVED** [review]
Thanks!

**Pull request merged**